src/library_manager/lib_manager.c

```diff
@@ -563,69 +567,110 @@ int lib_manager_load_library(uint32_t dma_id, uint32_t lib_id)
+		return 0;
 
-	lib_manager_init();
+	dma_ext = rzalloc(SOF_MEM_ZONE_RUNTIME, 0, SOF_MEM_CAPS_RAM,
+			  sizeof(*dma_ext));
+	if (!dma_ext)
+		return -ENOMEM;
 
-	ret = lib_manager_dma_init(&dma_ext, dma_id);
+	ret = lib_manager_dma_init(dma_ext, dma_id);
 	if (ret < 0)
-		return ret;
+		goto err_dma_init;
 
 #if CONFIG_ZEPHYR_NATIVE_DRIVERS
-	ret = dma_get_attribute(dma_ext.dma->z_dev, DMA_ATTR_BUFFER_ADDRESS_ALIGNMENT,
-				&addr_align);
+	ret = dma_get_attribute(dma_ext->dma->z_dev, DMA_ATTR_BUFFER_ADDRESS_ALIGNMENT,
+				&dma_ext->addr_align);
 #else
-	ret = dma_get_attribute_legacy(dma_ext.dma, DMA_ATTR_BUFFER_ADDRESS_ALIGNMENT,
-				       &addr_align);
+	ret = dma_get_attribute_legacy(dma_ext->dma, DMA_ATTR_BUFFER_ADDRESS_ALIGNMENT,
+				       &dma_ext->addr_align);
 #endif
 	if (ret < 0)
-		goto e_init;
+		goto err_dma_init;
 
-	/* allocate temporary manifest buffer */
-	man_tmp_buffer = (__sparse_force void __sparse_cache *)rballoc_align(0, SOF_MEM_CAPS_DMA,
-								MAN_MAX_SIZE_V1_8, addr_align);
-	if (!man_tmp_buffer) {
-		ret = -ENOMEM;
-		goto e_init;
-	}
-
-	ret = lib_manager_dma_buffer_alloc(&dma_ext, MAN_MAX_SIZE_V1_8, addr_align);
+	ret = lib_manager_dma_buffer_alloc(dma_ext, MAN_MAX_SIZE_V1_8);
 	if (ret < 0)
-		goto e_buf;
+		goto err_dma_buffer;
+
+	dma_block_cfg.dest_address = dma_ext->dma_addr;
 
 	ret = core_kcps_adjust(cpu_get_id(), CLK_MAX_CPU_HZ / 1000);
 	if (ret < 0)
-		goto cleanup;
-
-	dma_block_cfg.dest_address = dma_ext.dma_addr;
+		goto err_dma_buffer;
 
-	ret = dma_config(dma_ext.chan->dma->z_dev, dma_ext.chan->index, &config);
+	ret = dma_config(dma_ext->chan->dma->z_dev, dma_ext->chan->index, &config);
 	if (ret < 0)
-		goto cleanup;
+		goto err_dma;
 
-	ret = dma_start(dma_ext.chan->dma->z_dev, dma_ext.chan->index);
+	ret = dma_start(dma_ext->chan->dma->z_dev, dma_ext->chan->index);
 	if (ret < 0)
-		goto cleanup;
+		goto err_dma;
+
+	_ext_lib->runtime_data = dma_ext;
+
+	return 0;
+
+err_dma:
+	core_kcps_adjust(cpu_get_id(), -(CLK_MAX_CPU_HZ / 1000));
+
+err_dma_buffer:
+	lib_manager_dma_deinit(dma_ext, dma_id);
+
+err_dma_init:
+	rfree(dma_ext);
+	return ret;
+}
+
+int lib_manager_load_library(uint32_t dma_id, uint32_t lib_id)
+{
+	void __sparse_cache *man_tmp_buffer;
+	struct lib_manager_dma_ext *dma_ext;
+	struct ext_library *_ext_lib;
+	int ret, ret2;
+
+	if (lib_id >= LIB_MANAGER_MAX_LIBS) {
+		tr_err(&lib_manager_tr,
+		       "lib_manager_load_library(): invalid lib_id: %u", lib_id);
+		return -EINVAL;
+	}
+
+	lib_manager_init();
+
+	_ext_lib = ext_lib_get();
+
+	if (lib_id == 0 || !_ext_lib->runtime_data) {
+		ret = lib_manager_setup(dma_id);
+		if (ret)
+			return ret;
+
+		if (lib_id == 0)
+			return 0;
+	}
+
+	dma_ext = _ext_lib->runtime_data;
+
+	/* allocate temporary manifest buffer */
+	man_tmp_buffer = (__sparse_force void __sparse_cache *)
+			rballoc_align(0, SOF_MEM_CAPS_DMA,
+				      MAN_MAX_SIZE_V1_8, dma_ext->addr_align);
+	if (!man_tmp_buffer) {
+		ret = -ENOMEM;
+		goto e_init;
```

Comment:
dont we have to free dma_ext and kcps_adjust in this case too?

Reply:
adjust kcps is done because of recreating flow from windows driver in FW. Please use IPC4_RAT_DSP_KCPS instead.

Reply:
I would not require another message (two actually) to be sent when the same thing can be achieved without that said message.